File: discovery-ai-agent/backend/app/models/discovery.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String, Text
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship


class Base(DeclarativeBase):
    pass


class ProjectStatus(str, enum.Enum):
    DRAFT = "DRAFT"
    IN_PROGRESS = "IN_PROGRESS"
    BT_READY = "BT_READY"
    APPROVED = "APPROVED"


class ProjectStage(str, enum.Enum):
    CONTEXT = "CONTEXT"
    PROBLEM = "PROBLEM"
    GOAL = "GOAL"
    BUSINESS_EFFECT = "BUSINESS_EFFECT"
    AS_IS = "AS_IS"
    TO_BE = "TO_BE"
    USE_CASES = "USE_CASES"
    REQUIREMENTS = "REQUIREMENTS"
    RISKS = "RISKS"
    FINAL_BT = "FINAL_BT"


class ArtifactType(str, enum.Enum):
    CONTEXT = "CONTEXT"
    PROBLEM = "PROBLEM"
    GOAL = "GOAL"
    BUSINESS_EFFECT = "BUSINESS_EFFECT"
    AS_IS = "AS_IS"
    TO_BE = "TO_BE"
    USE_CASES = "USE_CASES"
    FUNCTIONAL_REQUIREMENTS = "FUNCTIONAL_REQUIREMENTS"
    NON_FUNCTIONAL_REQUIREMENTS = "NON_FUNCTIONAL_REQUIREMENTS"
    RISKS = "RISKS"
    GLOSSARY = "GLOSSARY"
    FINAL_BT = "FINAL_BT"


class DiscoveryProject(Base):
    __tablename__ = "discovery_projects"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_name: Mapped[str] = mapped_column(String(255), nullable=False)
    business_domain: Mapped[str | None] = mapped_column(String(255), nullable=True)
    status: Mapped[ProjectStatus] = mapped_column(Enum(ProjectStatus), default=ProjectStatus.DRAFT, nullable=False)
    current_stage: Mapped[ProjectStage] = mapped_column(Enum(ProjectStage), default=ProjectStage.CONTEXT, nullable=False)
    jira_epic_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    artifacts: Mapped[list["DiscoveryArtifact"]] = relationship(back_populates="project", cascade="all, delete-orphan")


class DiscoveryArtifact(Base):
    __tablename__ = "discovery_artifacts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("discovery_projects.id", ondelete="CASCADE"), nullable=False)
    artifact_type: Mapped[ArtifactType] = mapped_column(Enum(ArtifactType), nullable=False)
    content: Mapped[str] = mapped_column(Text, default="", nullable=False)
    version: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    project: Mapped[DiscoveryProject] = relationship(back_populates="artifacts")

File: discovery-ai-agent/frontend/src/types/discovery.ts
export type ProjectStatus = 'DRAFT' | 'IN_PROGRESS' | 'BT_READY' | 'APPROVED'
export type ProjectStage = 'CONTEXT' | 'PROBLEM' | 'GOAL' | 'BUSINESS_EFFECT' | 'AS_IS' | 'TO_BE' | 'USE_CASES' | 'REQUIREMENTS' | 'RISKS' | 'FINAL_BT' | 'VALIDATION_REPORT'
export type ArtifactType = 'CONTEXT' | 'PROBLEM' | 'GOAL' | 'BUSINESS_EFFECT' | 'AS_IS' | 'TO_BE' | 'USE_CASES' | 'FUNCTIONAL_REQUIREMENTS' | 'NON_FUNCTIONAL_REQUIREMENTS' | 'RISKS' | 'GLOSSARY' | 'FINAL_BT' | 'VALIDATION_REPORT'

export interface Project { id: string; project_name: string; business_domain?: string; status: ProjectStatus; current_stage: ProjectStage; jira_epic_url?: string }
export interface Artifact { id: string; project_id: string; artifact_type: ArtifactType; content: string; structured_content?: Record<string, unknown>; version: number }

export type ContextReadinessStatus = 'ready' | 'warning' | 'blocked'
export interface ContextCoverage {
  manual_context: boolean
  documents: boolean
  systems: boolean
  processes: boolean
  roles: boolean
  integrations: boolean
  bpmn: boolean
  kpi: boolean
  sla: boolean
  constraints: boolean
}
export interface ContextReadiness {
  status: ContextReadinessStatus
  score: number
  can_go_to_problem: boolean
  summary: string
  blocking_reasons: string[]
  warnings: string[]
  next_actions: string[]
}

export type AssistantMessageRole = 'user' | 'assistant' | 'system' | 'tool'

export interface AssistantSession {
  id: string
  project_id: string
  title: string
  status: string
  created_at: string
  updated_at: string
}

export interface AssistantMessage {
  id: string
  project_id: string
  session_id: string
  role: AssistantMessageRole
  content: string
  payload?: Record<string, unknown> | null
  created_at: string
}

export interface AssistantAction {
  id: string
  project_id: string
  session_id: string
  message_id: string
  action_type: string
  target_artifact_type?: ArtifactType | null
  status: 'proposed' | 'applied' | 'rejected' | string
  proposed_patch?: Record<string, unknown>
  preview?: Record<string, unknown>
  result?: Record<string, unknown>
  metadata?: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface AssistantChatRequest {
  message: string
  session_id?: string | null
  artifact_type?: ArtifactType | null
  context?: Record<string, unknown> | null
}

export interface AssistantChatResponse {
  ok: boolean
  session_id: string
  user_message: AssistantMessage
  assistant_message: AssistantMessage
  intent?: Record<string, unknown>
  action?: AssistantAction | null
  preview?: Record<string, unknown> | null
  warnings?: string[]
  errors?: string[]
}

export interface AssistantSessionsResponse {
  ok: boolean
  sessions: AssistantSession[]
}

export interface AssistantMessagesResponse {
  ok: boolean
  messages: AssistantMessage[]
}

export interface AssistantApplyPatchRequest {
  session_id: string
  action_id: string
}

export interface AssistantApplyPatchResponse {
  ok: boolean
  artifact: Artifact
  action: AssistantAction
}
